server/container/guice/cassandra-rabbitmq-guice/src/test/java/org/apache/james/RabbitMQJamesServerReprocessingTest.java

```diff
@@ -0,0 +1,91 @@
+import org.apache.james.modules.protocols.SmtpGuiceProbe;
+import org.apache.james.utils.MailRepositoryProbeImpl;
+import org.apache.james.utils.SMTPMessageSender;
+import org.apache.james.utils.WebAdminGuiceProbe;
+import org.apache.james.webadmin.WebAdminConfiguration;
+import org.apache.james.webadmin.WebAdminUtils;
+import org.awaitility.Awaitility;
+import org.awaitility.Duration;
+import org.awaitility.core.ConditionFactory;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.restassured.RestAssured;
+import io.restassured.config.ParamConfig;
+import io.restassured.parsing.Parser;
+import io.restassured.specification.RequestSpecification;
+
+class RabbitMQJamesServerReprocessingTest {
+    private static final ConditionFactory AWAIT = Awaitility.await().atMost(Duration.ONE_MINUTE);
+    private static final MailRepositoryUrl SENDER_DENIED = MailRepositoryUrl.from("cassandra://var/mail/sender-denied/");
+    private RabbitMQExtension rabbitMQExtension = new RabbitMQExtension();
+    private RequestSpecification webAdminApi;
+
+    @RegisterExtension
+    JamesServerExtension jamesServerExtension = CassandraRabbitMQJamesServerFixture
+        .baseExtensionBuilder(rabbitMQExtension)
+        .server(configuration -> GuiceJamesServer
+            .forConfiguration(configuration)
+            .combineWith(CassandraRabbitMQJamesServerMain.MODULES)
+            .overrideWith(new TestJMAPServerModule(10))
+            .overrideWith(binder -> binder.bind(WebAdminConfiguration.class).toInstance(WebAdminConfiguration.TEST_CONFIGURATION))
+            .overrideWith(JmapJamesServerContract.DOMAIN_LIST_CONFIGURATION_MODULE))
+        .build();
+
+    @BeforeEach
+    void setUp(GuiceJamesServer server) {
+        RestAssured.defaultParser = Parser.JSON;
+        webAdminApi = WebAdminUtils.spec(server.getProbe(WebAdminGuiceProbe.class).getWebAdminPort())
+            .config(WebAdminUtils.defaultConfig()
+                .paramConfig(new ParamConfig(REPLACE, REPLACE, REPLACE)));
+    }
+
+    @Test
+    void reprocessingADeniedMailShouldNotLooseIt(GuiceJamesServer server) throws Exception {
+        new SMTPMessageSender("other.com")
+            .connect("127.0.0.1", server.getProbe(SmtpGuiceProbe.class).getSmtpPort())
+            .sendMessage("denied@other.com", "any@domain.tld");
+
+        MailRepositoryProbeImpl mailRepositoryProbe = server.getProbe(MailRepositoryProbeImpl.class);
+        AWAIT.until(() -> mailRepositoryProbe.listMailKeys(SENDER_DENIED).size() == 1);
+
+        String taskId = webAdminApi
+            .param("action", "reprocess")
+            .patch("/mailRepositories/var%2Fmail%2Fsender-denied/mails")
+            .jsonPath()
+            .get("taskId");
+
+        webAdminApi.get("/tasks/" + taskId + "/await");
+
+        AWAIT.until(() -> mailRepositoryProbe.listMailKeys(SENDER_DENIED).size() == 1);
```

Comment:
Good question!

Awaiting the task (l86) ensure the reprocessing started. Thus the repository is empty.
Of course, we can not assert that it is empty at that very moment (as you can not 100% ensure reprocessing is already completed). IMO awaiting the task is a strong enough guaranty that the mail was processed.

The second await is here to ensure the reprocessing completes.

Mail processing logs back the analysis up.

I understand this can be a bit tricky to understand thus I will add some comments about it.

Reply:
Why not doing a

```
         AWAIT.until(() -> mailRepositoryProbe.listMailKeys(SENDER_DENIED).size() == 0);
```

before the WebAdmin's `await`?

Reply:
Then you rely on saying that the check is performed after the mail is removed and before this is fully processed.

You have no guaranty that your poll interval is low enough. It could lead to instable test.

Reply:
Currently the test is structured as such that you have no guarantee that you enforce what you want to enforce, it could well be a noop.